Unify frontend monitoring

ifrost opened the pull request:
Just an idea to unify the way we push frontend logs/errors to Faro and make it a bit easier to debug locally (with additional debug level config). Also, this way all frontend logs will have `source` in the context so it will be easier to search and aggregate.

* `createMonitoringLogger(source)` creates a scoped logger that always adds `{ source }` to the context
* updated logging in alerting, dashboards, correlations and plugins
* to enable [Faro debugging](https://github.com/grafana/faro-web-sdk/blob/main/docs/sources/tutorials/quick-start-browser.md#how-to-activate-debugging) use the following config in custom.ini:

```
[log.frontend]
enabled = true
internal_logger_level = 4
```

ifrost committed:
Unify frontend monitoring

ifrost committed:
Add missing mock

ifrost committed:
Add missing mock

academo commented on public/app/features/plugins/sandbox/utils.ts:
```diff
@@ -24,35 +19,22 @@ export function assertNever(x: never): never {
   throw new Error(`Unexpected object: ${x}. This should never happen.`);
 }
 
+const sandboxLogger = createMonitoringLogger('features.plugins.sandbox', { monitorOnly: String(monitorOnly) });
```
All this feature monitoring is already based on the source being "sandbox" I prefer we keep it like that.

ifrost committed:
Keep source:sandbox

aangelisc commented on public/app/features/plugins/sql/components/configuration/useMigrateDatabaseFields.ts:
```diff
@@ -1,8 +1,9 @@
-import { logDebug, config } from '@grafana/runtime';
+import { config } from '@grafana/runtime';
 
+import { pluginsLogger } from '../../../utils';
```
I think this may cause issues for us when externalising the sql data sources, @gabor am I correct?

ifrost replied:
It still using `@grafana/runtime` to call `logDebug` so should not cause problems, but let me know.

gabor replied:
@ifrost sorry, we need to stay inside `public/app/features/plugins/sql` , this import goes to `public/app/features/plugins/utils.ts`.

ifrost replied:
I've created a separate logger here: 2fabfd1b504f122d62468d82bdf5b1eb2755cf45

gillesdemey commented on public/app/features/alerting/unified/Analytics.ts:
```diff
@@ -22,18 +21,14 @@ export const LogMessages = {
-  }
+const alertingLogger = createMonitoringLogger('features.alerting', { module: 'Alerting' });
+
+export function logInfo(message: string, context?: Record<string, string>) {
```
Should this be rename to `logAlertingInfo` since it uses the alerting logger specifically?

ifrost replied:
Up to you. I kept the original name to reduce number of changes (the logic is the same with additional `source` param sent in context).

gillesdemey replied:
I've appending some changes :)  https://github.com/grafana/grafana/pull/80075/commits/2dac8798806a1deaec91d6a83de6e6524f0b3770

gillesdemey reviewed (approved):
LGTM from alerting, just one small nit

gabor reviewed (changes requested):
the `public/app/plugins/sql` folder imports something from it's parent folder, unfortunately we cannot have that (because we are working on extracting the whole `public/app/plugins/sql` folder from the grafana-repo), so could you look for an alternative solution please?

ifrost committed:
Merge branch 'main' into ifrost/unify-frontend-error-monitoring

ifrost committed:
Create separate logger for plugins/sql package

ifrost commented:
> the `public/app/plugins/sql` folder imports something from it's parent folder, unfortunately we cannot have that (because we are working on extracting the whole `public/app/plugins/sql` folder from the grafana-repo), so could you look for an alternative solution please?

Sure thing. I've created a separate logger in 2fabfd1b504f122d62468d82bdf5b1eb2755cf45.

gabor reviewed (approved):
the changes in `public/app/features/sql` look fine 👍 (i did not test the PR by running it)

gillesdemey committed:
chore: rename "logAlertingError" to "logError"

harisrozajac reviewed (approved):
looks great 🚀

ifrost committed:
Merge branch 'main' into ifrost/unify-frontend-error-monitoring

ifrost commented:
@grafana/grafana-frontend-platform, @grafana/plugins-platform-frontend, @grafana/dashboards-squad - you are the owners of files affected by this PR. Could you please have a look and let me know if you're okay with proposed changes?

ivanortegaalba commented:
I'm kind of confused with the logging vs the Faro monitoring itself. AFAIK:
*`pushX` go to the browser logger and Faro takes them and sends them to the Faro collector
* Uncontrolled errors are reported as Errors to Faro

Are error logs reported as errors to Faro too? If I'm not wrong they are reported as `console` whatever that means. Just because I want to make sure we use logs vs pushing errors.

See this https://github.com/grafana/faro-web-sdk/tree/main/packages/web-sdk#example

I didn't have time today to go into details, but what I would expect is to have a single function to call when I catch an error and I want to report it.

Maybe you @ifrost went further here

ifrost committed:
Merge branch 'main' into ifrost/unify-frontend-error-monitoring

ifrost commented:
> I'm kind of confused with the logging vs the Faro monitoring itself. AFAIK: *`pushX` go to the browser logger and Faro takes them and sends them to the Faro collector
> 
> * Uncontrolled errors are reported as Errors to Faro

Yep, uncontrolled should be reported as errors (at least I don't see anything in the config that would change that behavior, but I can double check).

> 
> Are error logs reported as errors to Faro too? If I'm not wrong they are reported as `console` whatever that means. Just because I want to make sure we use logs vs pushing errors.
> See this https://github.com/grafana/faro-web-sdk/tree/main/packages/web-sdk#example
> 
> I didn't have time today to go into details, but what I would expect is to have a single function to call when I catch an error and I want to report it.

Yes, you can publish caught errors with a single function `logError` from `grafana/runtime`. I've added a wrapper to that and other `pushX` functions called `createMonitoringLogger` to ensure `source` is always passed when anything is pushed in controlled way (uncontrolled errors won't have it). 

But I also added additional logger for debugging (`{createLogger } from '@grafana/ui';`) to print to the browser log whatever is suppose be pushed to Faro for easier debugging and development (maybe there's a better way to do it with Faro's instrumentation?).

But, yeah, I wasn't aware we instrument `console.log` :/  which could lead to pushing logs twice now. I'll put it back to draft to double check it. Thanks @ivanortegaalba

ifrost committed:
Use internal Faro logging for debugging instead of redundant browser logging

ifrost commented:
@ivanortegaalba - I've changed it slightly. There's now just a simple wrapper for `pushX` methods to have a consistent `source` property passed + default context for each logger. For debugging it's now possible to enable internal logger in Faro with:

```
[log.frontend]
enabled = true
internal_logger_level = 4
```

ifrost committed:
Merge branch 'main' into ifrost/unify-frontend-error-monitoring

# Conflicts:
#	packages/grafana-sql/src/utils/logging.ts

ifrost committed:
Post-merge fix

ivanortegaalba commented on conf/defaults.ini:
```diff
@@ -1038,6 +1038,9 @@ instrumentations_console_enabled = false
 # Should webvitals instrumentation be enabled, only affects Grafana Javascript Agent
 instrumentations_webvitals_enabled = false
 
+# level of internal logging for debugging Grafana Javascript Agent. 0 = OFF, 4 = VERBOSE
```
Can only accept 0 or 4? I'd add all the specific values this option accepts. WDYT?

ifrost replied:
Updated in 80bed7a5a35b41efc191feaf22120d659340a3d9

ivanortegaalba commented on public/app/features/query/utils.ts:
```diff
@@ -0,0 +1,3 @@
+import { createMonitoringLogger } from '@grafana/runtime';
+
+export const queryLogger = createMonitoringLogger('features/query');
```
```suggestion
export const queryLogger = createMonitoringLogger('features.query');
```
Should we use `.` to be consistent with the rest?

ifrost replied:
Good catch! Fixed in d9258d12509c6b21bbd47451914e6b65745b5b88

ivanortegaalba commented on packages/grafana-runtime/src/utils/logging.ts:
```diff
@@ -57,3 +57,22 @@ export function logError(err: Error, contexts?: LogContext) {
   }
 }
+
+/**
+ * Wrapper for logX functions.
+ * Creates a logger for frontend monitoring which contains provided "source" and context as defaults when calling logX()
+ */
```
```suggestion
/**
 * Creates a monitoring logger with four levels of logging methods: `logDebug`, `logInfo`, `logWarning`, and `logError`.
 * These methods use `faro.api.pushX` web SDK methods to report these logs or errors to the Faro collector.
 *
 * @param {string} source - Identifier for the source of the log messages.
 * @param {LogContext} [defaultContext] - Context to be included in every log message.
 *
 * @returns {Object} Logger object with four methods:
 * - `logDebug(message: string, contexts?: LogContext)`: Logs a debug message.
 * - `logInfo(message: string, contexts?: LogContext)`: Logs an informational message.
 * - `logWarning(message: string, contexts?: LogContext)`: Logs a warning message.
 * - `logError(error: Error, contexts?: LogContext)`: Logs an error message.
 * Each method combines the `defaultContext` (if provided), the `source`, and an optional `LogContext` parameter into a full context that is included with the log message.
 */
```
I just generated some extra comments with Copilot, feel free to edit it :)

ivanortegaalba commented on packages/grafana-runtime/src/utils/logging.ts:
```diff
@@ -57,3 +57,22 @@ export function logError(err: Error, contexts?: LogContext) {
+  });
+
+  return {
+    logDebug: (message: string, contexts?: LogContext) => logDebug(message, createFullContext(contexts)),
+    logInfo: (message: string, contexts?: LogContext) => logInfo(message, createFullContext(contexts)),
+    logWarning: (message: string, contexts?: LogContext) => logWarning(message, createFullContext(contexts)),
+    logError: (error: Error, contexts?: LogContext) => logError(error, createFullContext(contexts)),
```
Add some comments to get IDE help when using the returned object since the original comments are lost.
```suggestion
    /**
     * Logs a debug message with optional additional context.
     * @param {string} message - The debug message to be logged.
     * @param {LogContext} [contexts] - Optional additional context to be included.
     */
    logDebug: (message: string, contexts?: LogContext) => logDebug(message, createFullContext(contexts)),

    /**
     * Logs an informational message with optional additional context.
     * @param {string} message - The informational message to be logged.
     * @param {LogContext} [contexts] - Optional additional context to be included.
     */
    logInfo: (message: string, contexts?: LogContext) => logInfo(message, createFullContext(contexts)),

    /**
     * Logs a warning message with optional additional context.
     * @param {string} message - The warning message to be logged.
     * @param {LogContext} [contexts] - Optional additional context to be included.
     */
    logWarning: (message: string, contexts?: LogContext) => logWarning(message, createFullContext(contexts)),

    /**
     * Logs an error with optional additional context.
     * @param {Error} error - The error object to be logged.
     * @param {LogContext} [contexts] - Optional additional context to be included.
     */
    logError: (error: Error, contexts?: LogContext) => logError(error, createFullContext(contexts)),
```

ivanortegaalba reviewed (approved):
I'm ok with all these changes, I think it is good progress on using Faro in the different parts of the product. Good initiative @ifrost!

ifrost committed:
Merge branch 'main' into ifrost/unify-frontend-error-monitoring

ifrost commented on public/app/core/services/echo/backends/grafana-javascript-agent/GrafanaJavascriptAgentBackend.ts:
```diff
@@ -74,6 +74,7 @@ export class GrafanaJavascriptAgentBackend
       batching: {
         sendTimeout: 1000,
       },
+      internalLoggerLevel: options.internalLoggerLevel || defaultInternalLoggerLevel,
```
It's official part of SDK ([more docs here](https://github.com/grafana/faro-web-sdk/blob/main/docs/sources/tutorials/quick-start-browser.md#how-to-activate-debugging)). It doesn't change the behavior but allows additional debug messages on what's going on under the hood (e.g. allows you to see when logs were sent out, or if a log was skipped because it's a duplicate).

ifrost committed:
Add more docs about debug levels

ifrost committed:
Unify logger names

ifrost committed:
Update packages/grafana-runtime/src/utils/logging.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>

ifrost committed:
Update packages/grafana-runtime/src/utils/logging.ts

Co-authored-by: Ivan Ortega Alba <ivanortegaalba@gmail.com>
